truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedTruffleRuntime.java

```diff
@@ -912,10 +913,45 @@ private void notifyCompilationFailure(OptimizedCallTarget callTarget, Throwable
     protected void onEngineCreated(EngineData engine) {
     }
 
+    private long stoppedCompilationTime = 0;
+    private boolean logShutdownCompilations = true;
```

Comment:
submitForCompilation is not thread-safe. We need to handle these fields to be modified from multpile threads. So we would need to synchronize accessing these fields.

Reply:
In this case synchronization is not really required for correctness (the worst thing that can happen is that we schedule a compilation too much if compilations are stopped or we log more than once if compilation has been shutdown) but I've now changed both fields to `AtomicLong` and `AtomicBoolean` to also make the code formally correct.